From 70fbfa2da064460e9dade676a2dce4d44ba28793 Mon Sep 17 00:00:00 2001
From: Tim Besard <tim.besard@gmail.com>
Date: Mon, 7 Jul 2025 15:08:44 +0200
Subject: [PATCH] Generate bytecode instead of native binaries.

This makes it easier to run binaries during cross compilation.
For Windows, we need to explicitly call `ocamlrun`.
---
 lib/pack/dune             | 3 ++-
 src/stage1/dune           | 1 +
 src/stage2/dune           | 5 +++--
 src/stage3/anonymize/dune | 1 +
 src/stage3/dune           | 9 +++++----
 5 files changed, 12 insertions(+), 7 deletions(-)

diff --git a/lib/pack/dune b/lib/pack/dune
index 06e53983..bdc9776c 100644
--- a/lib/pack/dune
+++ b/lib/pack/dune
@@ -4,12 +4,13 @@
 (executable
   (name pack)
   (modules pack)
+  (modes byte)
 )
 
 (rule
   (targets menhirLib.ml menhirLib.mli)
   (deps (glob_files ../*.{ml,mli}) menhirLib.mlpack)
-  (action (run ./pack.exe))
+  (action (run ocamlrun %{dep:./pack.bc}))
 )
 
 ;; We can then compile MenhirLib from menhirLib.{ml,mli}
diff --git a/src/stage1/dune b/src/stage1/dune
index 3a733d31..bd871103 100644
--- a/src/stage1/dune
+++ b/src/stage1/dune
@@ -13,6 +13,7 @@
 
 (executable
   (name main)
+  (modes byte)
   (libraries
     unix
     vendored_fix
diff --git a/src/stage2/dune b/src/stage2/dune
index 49dc0ec0..2776b4f9 100644
--- a/src/stage2/dune
+++ b/src/stage2/dune
@@ -29,7 +29,7 @@
 ;; developer's machine.
 
 (env
-  (_ (binaries ../stage1/main.exe (../stage1/main.exe as menhir)))
+  (_ (binaries ../stage1/main.bc (../stage1/main.bc as menhir)))
 )
 
 ;; Menhir's parser is generated by Menhir.
@@ -62,6 +62,7 @@
 
 (executable
   (name main)
+  (modes byte)
   (libraries
     unix
     vendored_fix
@@ -84,7 +85,7 @@
 (install
   (section bin)
   (package menhir)
-  (files (./main.exe as menhir))
+  (files (./main.bc as menhir))
 )
 
 ;; -----------------------------------------------------------------------------
diff --git a/src/stage3/anonymize/dune b/src/stage3/anonymize/dune
index e50578fe..e78fd4e7 100644
--- a/src/stage3/anonymize/dune
+++ b/src/stage3/anonymize/dune
@@ -1,4 +1,5 @@
 (executable
   (name anonymize)
+  (modes byte)
   (libraries str)
 )
diff --git a/src/stage3/dune b/src/stage3/dune
index 51df40e6..21929cd8 100644
--- a/src/stage3/dune
+++ b/src/stage3/dune
@@ -23,10 +23,10 @@
 
 ;; -----------------------------------------------------------------------------
 
-;; Bind the name "menhir" to "../stage2/main.exe" within the present scope.
+;; Bind the name "menhir" to "../stage2/main.bc" within the present scope.
 
 (env
-  (_ (binaries ../stage2/main.exe (../stage2/main.exe as menhir)))
+  (_ (binaries ../stage2/main.bc (../stage2/main.bc as menhir)))
 )
 
 ;; Menhir's parser is generated by Menhir.
@@ -50,6 +50,7 @@
 
 (executable
   (name main)
+  (modes byte)
   (libraries
     unix
     vendored_fix
@@ -79,13 +80,13 @@
 
 (rule
   (with-stdout-to parser.stage2.ml
-    (run anonymize/anonymize.exe %{dep:../stage2/parser.ml})
+    (run ocamlrun anonymize/anonymize.bc %{dep:../stage2/parser.ml})
   )
 )
 
 (rule
   (with-stdout-to parser.stage3.ml
-    (run anonymize/anonymize.exe %{dep:parser.ml})
+    (run ocamlrun anonymize/anonymize.bc %{dep:parser.ml})
   )
 )
 
-- 
2.50.0

